python/ray/dag/compiled_dag_node.py

```diff
@@ -867,6 +876,81 @@ def _get_node_id(self):
 
         self._dag_submitter.start()
         self._dag_output_fetcher.start()
+        self._max_concurrent_executions = self._compute_max_concurrent_executions()
+
+    @property
+    def max_concurrent_executions(self) -> int:
+        """
+        Get the maximum number of concurrent executions that the DAG
+        can perform without blocking.
+        """
+        return self._max_concurrent_executions
+
+    def _compute_max_concurrent_executions(self):
+        """
+        Compute the max possible concurrent executions of the DAG. This is the
+        number of inputs the DAG could start processing without blocking when
+        none of the outputs is consumed.
+
+        Returns:
+            The max possible concurrent executions of the DAG.
+
+        NOTE: The max possible concurrent executions is determined by the
+        "capacity" of the DAG, i.e., the number of intermediate results the
+        DAG can hold. The DAG holds these intermediate results in either
+        the intermediate channels or the actors, specifically:
+        - Channel capacity: either 1 or 0 depending on the channel type.
+        - Actor capacity: capacity is 1 because the actor can fetch and hold one
+            object from each input channel during execution. For an actor with
+            multiple methods, the capacity is still 1 because the current method
+            needs to finish (which flushes the held object) before the next method
+            can start.
+
+        This method performs a modified Dijkstra's algorithm to compute the
+        capacity of the DAG.
+        """
+        capa = {idx: float("inf") for idx in self.idx_to_task}
+        priority_queue = []
+
+        # Initial capacity is 1 as input channel has capacity 1.
+        capa[self.input_task_idx] = 1
+        heapq.heappush(priority_queue, (1, self.input_task_idx, set()))
+
+        while priority_queue:
+            cur_capa, cur_idx, cur_actors = heapq.heappop(priority_queue)
+            if cur_idx == self.output_task_idx:
+                return cur_capa
+
+            chan_capa = self.idx_to_task[cur_idx].output_channel.capacity()
+
+            for downstream_idx, actor in self.idx_to_task[
+                cur_idx
+            ].downstream_node_idxs.items():
+                if actor is not None and actor not in cur_actors:
+                    # If the actor is not used in the current path, its method
+                    # can be executed, therefore 1 object can be buffered at
+                    # this actor, and chan_capa can be buffered at the output channel.
+                    actor_capa = 1
+                    downstream_capa = cur_capa + actor_capa + chan_capa
+                else:
```

Comment:
```
driver -> a.f0 -> b.f -> a.f1 -> driver
```
* In the above example, there are two actors, `a` and `b`, and all channels are shared memory channel. 
* The test `test_composite_channel_two_actors` uses the same example, and the capacity is 5.

However, the capacity seems to be 3. To elaborate,

* t=0: ch0
* t=1: ch0, `a.f0`
* t=2: ch0, ch1 (`a.f0` finishes, so actor `a` starts to execute `a.f1`.)
* t=3: ch0, `b.f`
* t=4: ch0, ch2
* t=5: ch0, `a.f1`
* t=6: ch0, ch3 (`a.f1` finishes, so actor `a` starts to execute `a.f0`.)
* t=7: ch0, `a.f0`, ch3
* t=8: ch0, ch1, ch3
* t=9: ch0, `b.f`, ch3
* t=10: ch0, ch2, ch3
* t=11: ch0, `a.f1`, ch3 (`a.f1` will stuck there because it can't write into ch3.)

Do I miss anything?

Reply:
I think `b.f` and its output channel should be able to hold two other objects.
I had an illustration in the test:
        # driver -(5)-> a.inc(x) -(x)-> b.inc(4) -(3)-> a.inc(2) -(1)-> driver
We can also chat more offline.